### PR TITLE
WIP : Strassen algorithm for matrix multiplication

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

#15421
#16230

#### Brief description of what is fixed or changed

There had been numerous discussions about adding strassen algorithm.
Well, my idea here is about extending the `multiply` method as an algorithm dump,
and operators like `__mul__` can be redesigned, to incorporate some smart strategy, like switching the methods according to the matrix sizes.

#### Release Notes

- matrices
  - Added `method` option for `multiply`, where user can select matrix multiplication algorithms.
  - Added strassen algorithm for `multiply`

### CODE REVIEW

File: sympy/matrices/common.py [hunk 1]
```diff
@@ -2205,6 +2205,113 @@ def __sub__(self, a):
...
+            cls = mat.__class__
+            if i_new > i_old:
+                diff = i_new - i_old
+                mat = cls.vstack(mat, cls.zeros(diff, j_old))
+            if j_new > j_old:
+                diff = j_new - j_old
+                mat = cls.hstack(mat, cls.zeros(i_new, diff))
+            return mat
+
+        _strasson = MatrixArithmetic._eval_matrix_mul_strassen
```
Comment (oscargus) [inline sympy/matrices/common.py]: `_strassen`?

File: sympy/matrices/common.py [hunk 2]
```diff
@@ -2205,6 +2205,113 @@ def __sub__(self, a):
...
+
+        other : Matrix or Scalar
+            Matrix dimensions should match.
+
+            If ``other`` is scalar, it is equivalent of multiplying
+            `A` and `B I` where `I` is an identity matrix which has rows
+            and columns as same as `A` 's columns.
+
+        method : String, optional
+            Specifies the method used for mathematrix multiplication
```
Comment (oscargus) [inline sympy/matrices/common.py]: matrix

Good to mention the options.

**Comment by oscargus** [comment]
Is is more efficient to pad as one goes along compared to going for a power of two directly?

**Comment by oscarbenjamin** [comment]
Apart from the comments @oscargus already made this looks like a good change.

Do you have any timings? How big does the matrix need to be for this to help (for 1. float matrix, 2. numeric matrix, 3. symbolic matrix)?

File: sympy/matrices/common.py [hunk 3]
```diff
@@ -2205,6 +2205,113 @@ def __sub__(self, a):
...
+                diff = i_new - i_old
+                mat = cls.vstack(mat, cls.zeros(diff, j_old))
+            if j_new > j_old:
+                diff = j_new - j_old
+                mat = cls.hstack(mat, cls.zeros(i_new, diff))
+            return mat
+
+        _strasson = MatrixArithmetic._eval_matrix_mul_strassen
+
+        if (a_rows, a_cols) == (1, 1) and (b_rows, b_cols) == (1, 1):
```
Comment (oscarbenjamin) [inline sympy/matrices/common.py]: The Strassen algorithm wouldn't normally be used all the way down to 1x1. Maybe there should be an argument to specify the minimum size while recursing.

**Comment by oscarbenjamin** [comment]
This seems to be slower than naive multiplication for all sizes I've tested:
```
$ cat t.py 
from sympy import Matrix
import random
def randmat(n):
    return Matrix([[random.randint(1, 10) for _ in range(n)] for _ in range(n)])
$ python -m timeit -s 'from t import randmat' 'A=randmat(2); B=randmat(2); C=A.multiply(B, "strassen")'
200 loops, best of 5: 996 usec per loop
$ python -m timeit -s 'from t import randmat' 'A=randmat(2); B=randmat(2); C=A.multiply(B, "naive")'
2000 loops, best of 5: 138 usec per loop
```
Then size 10:
```
$ python -m timeit -s 'from t import randmat' 'A=randmat(10); B=randmat(10); C=A.multiply(B, "strassen")'
1 loop, best of 5: 495 msec per loop
$ python -m timeit -s 'from t import randmat' 'A=randmat(10); B=randmat(10); C=A.multiply(B, "naive")'
50 loops, best of 5: 7.5 msec per loop
```
At 64 I get:
```
$ python -m timeit -s 'from t import randmat' 'A=randmat(64); B=randmat(64); C=A.multiply(B, "naive")'
1 loop, best of 5: 1.68 sec per loop
$ python -m timeit -s 'from t import randmat' 'A=randmat(64); B=randmat(64); C=A.multiply(B, "strassen")'
1 loop, best of 5: 21.7 sec per loop
```

I think this is due to using Strassen all the way down to size 1x1. Instead it should be used down to maybe 16x16 or something.

File: sympy/matrices/common.py [hunk 4]
```diff
@@ -2205,6 +2205,183 @@ def __sub__(self, a):
...
+            a12 = self[:a_rows//2, -a_cols//2:]
+            a21 = self[-a_rows//2:, :a_cols//2]
+            a22 = self[-a_rows//2:, -a_cols//2:]
+
+            b11 = other[:b_rows//2, :b_cols//2]
+            b12 = other[:b_rows//2, -b_cols//2:]
+            b21 = other[-b_rows//2:, :b_cols//2]
+            b22 = other[-b_rows//2:, -b_cols//2:]
+
+            m1 = _strassen(a11 + a22, b11 + b22)
```
Comment (oscarbenjamin) [inline sympy/matrices/common.py]: The breakpoint argument needs to passed recursively here.

File: sympy/matrices/common.py [hunk 5]
```diff
@@ -2205,6 +2205,183 @@ def __sub__(self, a):
...
+        return ret[:a_rows, :b_cols]
+
+
+    def _eval_matrix_mul_strassen(self, other, breakpoint=(1, 1)):
+        """Strassen subroutine for matrix multiplication.
+
+        Parameters
+        ==========
+
+        breakpoint: tuple of two integers
```
Comment (oscarbenjamin) [inline sympy/matrices/common.py]: I don't think breakpoint is a good name for this. It should perhaps be minsize or something

**Comment by oscarbenjamin** [comment]
Testing with breakpoint = (16,16) and different size input matrices (all powers of 2) I find that this is not faster for moderately sized matrices:
```console
$ python -m timeit -s 'from t import randmat' 'A=randmat(2); B=randmat(2); C=A._eval_matrix_mul_strassen(B, (16, 16))'
2000 loops, best of 5: 180 usec per loop
$ python -m timeit -s 'from t import randmat' 'A=randmat(2); B=randmat(2); C=A._eval_matrix_mul(B)'
2000 loops, best of 5: 137 usec per loop
```
The examples above are for 2x2 matrices which should be similar since breakpoint is (16,16) but the Strassen method is slower because of the additional function call overhead in redirecting back to naive multiplication (presumably).

For other size matrices and breakpoint=(16,16) I get:
```
Size, Naive, Strassen
2x2 : 137 usec, 180 usec
64x64: 1.66 sec, 1.87 sec
128x128: 13.3 sec, 14.3 sec
256x256: ?
```

Timeit is slow here because it loops through 5 times so I've made a timing script (`t.py`):
```python
#!/usr/bin/env python

from sympy import Matrix
import random
import time
import sys

def randmat(n):
    return Matrix([[random.randint(1, 10) for _ in range(n)] for _ in range(n)])

n, b = sys.argv[1:]
n = int(n)
b = int(b)

print('%sx%s with minsize %sx%s' % (n, n, b, b))

def test_strassen():
    A = randmat(n)
    B = randmat(n)

    t = time.time()
    C = C=A._eval_matrix_mul_strassen(B, (b, b))
    return time.time() - t

def test_naive():
    A = randmat(n)
    B = randmat(n)

    t = time.time()
    C = C=A._eval_matrix_mul(B)
    return time.time() - t

# Time results:
print('naive    %5.2f seconds' % test_naive())
print('strassen %5.2f seconds' % test_strassen())
```

Now we still can't see performance improvement at 256x256:
```console
$ ./t.py 64 16
64x64 with minsize 16x16
naive     1.68 seconds
strassen  1.93 seconds
$ ./t.py 128 16
128x128 with minsize 16x16
naive    14.60 seconds
strassen 14.55 seconds
$ ./t.py 256 16
256x256 with minsize 16x16
naive    108.79 seconds
strassen 106.03 seconds
```

I note that Wikipedia suggests using naive multiplication for sizes as large as 25x25. Since Strassen here is slower for 16x16 we probably should use a larger minsize so:
```console
$ ./t.py 256 16
256x256 with minsize 16x16
naive    108.79 seconds
strassen 106.03 seconds
$ ./t.py 256 32
256x256 with minsize 32x32
naive    113.72 seconds
strassen 94.52 seconds
$ ./t.py 256 64
256x256 with minsize 64x64
naive    117.03 seconds
strassen 89.56 seconds
$ ./t.py 256 128
256x256 with minsize 128x128
naive    109.92 seconds
strassen 90.52 seconds
```
Although this method of timing clearly has a substantial standard deviation we can see that Strassen is clearly a bit faster the naive with a minsize of 64x64 which is possibly optimal for the implementation here (and for this kind of matrix of small integers).

On the other hand profiling Strassen with that (comment out call to test_naive):
```console
$ python -m cProfile -s cumulative t.py 256 64
256x256 with minsize 64x64
strassen 136.69 seconds
         247716053 function calls (235792971 primitive calls) in 138.236 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    521/1    0.008    0.000  138.236  138.236 {built-in method builtins.exec}
        1    0.003    0.003  138.236  138.236 t.py:3(<module>)
        1    0.006    0.006  137.420  137.420 t.py:17(test_strassen)
     57/1    0.012    0.000  136.679  136.679 common.py:2265(_eval_matrix_mul_strassen)
     56/7    0.016    0.000  131.466   18.781 common.py:2306(<lambda>)
       49    0.000    0.000  122.254    2.495 common.py:2354(multiply)
       49    3.471    0.071  122.254    2.495 dense.py:178(_eval_matrix_mul)
27414109/15520630   15.635    0.000   92.286    0.000 cache.py:92(wrapper)
200707/200706    4.979    0.000   77.197    0.000 operations.py:28(__new__)
   200705   11.206    0.000   45.348    0.000 add.py:78(flatten)
 12696822   16.019    0.000   34.455    0.000 numbers.py:1997(__add__)
 13045760    8.835    0.000   33.550    0.000 dense.py:200(<genexpr>)
13577846/13577834    6.987    0.000   26.277    0.000 sympify.py:391(_sympify)
 12845060   15.608    0.000   24.716    0.000 numbers.py:2038(__mul__)
14796786/14794194   13.893    0.000   21.821    0.000 sympify.py:78(sympify)
      331    0.008    0.000   11.605    0.035 dense.py:435(_new)
       26    0.099    0.004   11.597    0.446 matrices.py:2213(_handle_creation_inputs)
       24    0.000    0.000   11.245    0.469 {built-in method _functools.reduce}
```
With minsize 16 we get:
```console
256x256 with minsize 16x16
strassen 179.05 seconds
         280787908 function calls (271618224 primitive calls) in 180.661 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    521/1    0.009    0.000  180.662  180.662 {built-in method builtins.exec}
        1    0.003    0.003  180.662  180.662 t.py:3(<module>)
        1    0.007    0.007  179.794  179.794 t.py:17(test_strassen)
   2801/1    0.098    0.000  179.040  179.040 common.py:2265(_eval_matrix_mul_strassen)
   2800/7    0.091    0.000  173.723   24.818 common.py:2306(<lambda>)
     2401    0.006    0.000  117.003    0.049 common.py:2354(multiply)
     2401    4.439    0.002  116.997    0.049 dense.py:178(_eval_matrix_mul)
29256493/20135646   17.513    0.000   96.302    0.000 cache.py:92(wrapper)
614659/614658    7.273    0.000   75.552    0.000 operations.py:28(__new__)
    16403    0.059    0.000   48.109    0.003 dense.py:435(_new)
     1202    0.174    0.000   48.037    0.040 matrices.py:2213(_handle_creation_inputs)
     1200    0.002    0.000   47.713    0.040 {built-in method _functools.reduce}
    30528    3.088    0.000   47.543    0.002 matrices.py:2338(<listcomp>)
   614657   11.836    0.000   42.945    0.000 add.py:78(flatten)
 12407902   16.594    0.000   35.322    0.000 numbers.py:1997(__add__)
```
So it looks like the minsize 16 could be made faster if `_handle_creation_inputs` was faster. Both seem to spend a significant amount of time in `flatten`.

For reference numpy on this machine gives:
```julia
In [14]: def test_numpy(): 
    ...:     A = np.random.randint(10, size=(256, 256)) 
    ...:     B = np.random.randint(10, size=(256, 256))  
    ...:     C = A @ B 
    ...:                                                                                                                                                      

In [15]: %timeit test_numpy()                                                                                                                                 
71.1 ms ± 308 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
So numpy is 1000 times faster than either of these methods at size 256x256.

**Comment by czgdp1807** [comment]
@sylee957 Any news on this?

**Comment by sylee957** [comment]
Unfortunately, I lost some interest in making how to make good use of Strassen to multiply the matrix faster.
I'd defer to #15421

**Pull request closed without merging**